components/concept/engine/src/main/java/mozilla/components/concept/engine/InputResultDetail.kt

```diff
@@ -0,0 +1,267 @@
+
+/**
+ * The the content has no scrollable element.
+ *
+ * @see [InputResultDetail.isTouchUnhandled]
+ */
+@VisibleForTesting
+internal const val INPUT_UNHANDLED = 0
+
+/**
+ * The touch event is consumed by the [EngineView]
+ *
+ * @see [InputResultDetail.isTouchHandledByBrowser]
+ */
+@VisibleForTesting
+internal const val INPUT_HANDLED = 1
+
+/**
+ * The touch event is consumed by the website through it's own touch listeners.
+ *
+ * @see [InputResultDetail.isTouchHandledByWebsite]
+ */
+@VisibleForTesting
+internal const val INPUT_HANDLED_CONTENT = 2
+
+/**
+ * The website content is not scrollable.
+ */
+@VisibleForTesting
+internal const val SCROLL_DIRECTIONS_NONE = 0
+
+/**
+ * The website content can be scrolled to the top.
+ *
+ * @see [InputResultDetail.canScrollToTop]
+ */
+@VisibleForTesting
+internal const val SCROLL_DIRECTIONS_TOP = 1 shl 0
+
+/**
+ * The website content can be scrolled to the right.
+ *
+ * @see [InputResultDetail.canScrollToRight]
+ */
+@VisibleForTesting
+internal const val SCROLL_DIRECTIONS_RIGHT = 1 shl 1
+
+/**
+ * The website content can be scrolled to the bottom.
+ *
+ * @see [InputResultDetail.canScrollToBottom]
+ */
+@VisibleForTesting
+internal const val SCROLL_DIRECTIONS_BOTTOM = 1 shl 2
+
+/**
+ * The website content can be scrolled to the left.
+ *
+ * @see [InputResultDetail.canScrollToLeft]
+ */
+@VisibleForTesting
+internal const val SCROLL_DIRECTIONS_LEFT = 1 shl 3
+
+/**
+ * The website content cannot be overscrolled.
+ */
+@VisibleForTesting
+internal const val OVERSCROLL_DIRECTIONS_NONE = 0
+
+/**
+ * The website content can be overscrolled horizontally.
+ *
+ * @see [InputResultDetail.canOverscrollRight]
+ * @see [InputResultDetail.canOverscrollLeft]
+ */
+@VisibleForTesting
+internal const val OVERSCROLL_DIRECTIONS_HORIZONTAL = 1 shl 0
+
+/**
+ * The website content can be overscrolled vertically.
+ *
+ * @see [InputResultDetail.canOverscrollTop]
+ * @see [InputResultDetail.canOverscrollBottom]
+ */
+@VisibleForTesting
+internal const val OVERSCROLL_DIRECTIONS_VERTICAL = 1 shl 1
+
+/**
+ * All data about how a touch will be handled by the browser.
+ * - whether the event is used for panning/zooming by the browser / by the website or will be ignored.
+ * - whether the event can scroll the page and in what direction.
+ * - whether the event can overscroll the page and in what direction.
+ *
+ * Data wrapped by this class is safe to use after first calling [update] to update the default values.
+ *
+ * This class can be reused by calling [reset] to return all data values to it's default value.
+ */
+@Suppress("TooManyFunctions")
+class InputResultDetail {
+    /**
+     * Indicates who will use the current [MotionEvent].
+     *
+     * @see INPUT_UNHANDLED
+     * @see INPUT_HANDLED
+     * @see INPUT_HANDLED_CONTENT
+     */
+    @VisibleForTesting
+    internal var inputResult: Int = INPUT_UNHANDLED
```

Comment:
Is there any specific reason why we should have mutable values instead of using immutable ones?  Like a [data class and using copy](https://kotlinlang.org/docs/data-classes.html), the common pattern across AC is using immutable with data classes.

Reply:
As immutability brings multiple [benefits](https://hackernoon.com/5-benefits-of-immutable-objects-worth-considering-for-your-next-project-f98e7e85b6ac) with it.

Reply:
I had the wrong impression that this data would be updated a lot of times and wanted to avoid creating many new instances of that data class. Saw after this is not the case but forgot to change to a data class. Will do now. Thank you!

Reply:
Ohh. Using a data class and `copy` would not allow filtering what values this accepts like it's possible now with the `update` method.

Reply:
@Amejia481 Do you feel strongly about this being a data class?
I would've preferred this to encapsulate all the behavior but with it exposing the constants and putting the responsibility of checking the input data on users this will be just a shell that is harder to work with.

Reply:
Was thinking about morphing the current update method in a `copy` one to mimic the one from the data class while also being able to do the filtering.

Reply:
> @Amejia481 Do you feel strongly about this being a data class?

No really but they come with a lot of [goodness on them](https://kotlinlang.org/docs/data-classes.html). What concerns me the most it's having mutable values (`var`s) without a specific need, as mutable values could hide bugs that make life harder in the future, like making difficult to find concurrency issues.

Reply:
Will look into implementing the same methods as a data class to have those nice things and also ensure the copy method will do the needed filtering.

Reply:
Refactored `InputResultDetail` to follow the general data class API.
The `copy` metho will help filter unrecognized values and keep the functionality sane.
The `newInstance` method is the only way to create .. a new instance. With it using a boolean it's easy to control whether on GV the overscroll should start as enabled and for WV it should be disabled without exposing more of the constants of this class.